modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -117,13 +117,33 @@ func (suite *KeeperTestSuite) TestChanUpgradeInit() {
 			)
 
 			if tc.expPass {
-				suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.WriteUpgradeInitChannel(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, proposedUpgrade)
+				ctx := suite.chainA.GetContext()
+				suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.WriteUpgradeInitChannel(ctx, path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, proposedUpgrade)
 				channel := path.EndpointA.GetChannel()
 
+				events := ctx.EventManager().Events()
+				expEvents := ibctesting.EventsMap{
+					types.EventTypeChannelUpgradeInit: {
+						types.AttributeKeyPortID:                    path.EndpointA.ChannelConfig.PortID,
+						types.AttributeKeyChannelID:                 path.EndpointA.ChannelID,
+						types.AttributeCounterpartyPortID:           path.EndpointB.ChannelConfig.PortID,
+						types.AttributeCounterpartyChannelID:        path.EndpointB.ChannelID,
+						types.AttributeKeyUpgradeConnectionHops:     upgrade.Fields.ConnectionHops[0],
+						types.AttributeKeyUpgradeVersion:            upgrade.Fields.Version,
+						types.AttributeKeyUpgradeOrdering:           upgrade.Fields.Ordering.String(),
+						types.AttributeKeyUpgradeSequence:           fmt.Sprintf("%d", channel.UpgradeSequence),
+						types.AttributeKeyUpgradeChannelFlushStatus: channel.FlushStatus.String(),
+					},
+					sdk.EventTypeMessage: {
+						sdk.AttributeKeyModule: types.AttributeValueCategory,
+					},
+				}
```

Comment:
This doesn't seem to be a super clean looking test, I'm open to suggestions for cleaning this up.

Maybe we could create a function that returns the expected events for `OnChanUpgradeInit` and re-use that both in this test and where the events are emitted.

This might look a little cleaner, but the current version is more explicit.

Reply:
> Maybe we could create a function that returns the expected events for OnChanUpgradeInit and re-use that both in this test and where the events are emitted.

that does also sound better to me? Split the emitting from the creating in `events.go`. Addtional hurdle is adding all creating of events in `export_test` to have them be testable.

Reply:
The only issue I see of using here the function that creates the event is that we are using the code that we want to test to actually test it, so if in the future we remove by mistake an attribute, the test will still pass... The same thing as if we use the attribute key constants here: if at some point we changed by mistake the string, we will not catch that here...

Another thing to take into account is that with SDK v0.50 an extra attribute is appended to the events, so we need to take that one into account when doing the check. [This was my hacky way of quickly fixing it in the SDK v0.50 branch](https://github.com/cosmos/ibc-go/pull/3883/commits/72796c495d7368e4d5a36b760d92dd28ac13889e).

Reply:
ah, I thought we were testing for contents, not structure really. that's a valid point tho.

Reply:
Maybe we should be less strict with the `AssertEvents` fn in that case, and just verify that the provided events are a subset of the actual events, I figure as far as this test is concerned, we should only need to worry about the upgradability related events, and not any additional events that sdk emits

Or even 

```golang
sdk.EventTypeMessage: {
    sdk.AttributeKeyModule: types.AttributeValueCategory,
},
```

I feel like our current [check](https://github.com/cosmos/ibc-go/blob/dffc7b7d12f644fad4b8f1b8b882b35653322ed8/testing/events.go#L153) for length is actually going to be quite fragile, and break all event tests in the case of a new sdk message that we are not concerned with.

What do you think?

Reply:
The referenced length check should be fine because it references event attributes not event types, so a new sdk event shouldn't affect our events length of attributes. I believe we are searching a list of actual events for a subset of expected events.

I'm fine with how the code is currently situated. I agree it'd be nice to improve our approach to event testing so it feels natural. One problem we have already seen with the map approach is that it doesn't account for multiple emissions of the same event type

Reply:
okay sounds good to me, we can merge this one as is and iterate upon it later.